STAR --genomeDir [path_to_indexed_genome_reference_directory] --readFilesIn <trimmed_paired_end_fastq.gz file 1> <trimmed_paired_end_fastq.gz file 2> --readFilesCommand gzcat --runThreadN [desired_number_of_threads] --outFilterMultimapNmax 100 --outFilterIntronMotifs RemoveNoncanonicalUnannotated --alignEndsProtrude 10 ConcordantPair --outSAMtype BAM SortedByCoordinate --outFileNamePrefix [desired_prefix_for_output]